File: LFSR.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#define M (int) pow(2,24)
#define NUM_TOSS 10000
#define A 0x43FD43FD
#define C 0xC39EC3
unsigned int lfsr = 115;
unsigned int lcg = 115;
int findMode(char*);
void piEstimator(char*);
int numH(int, int*);
double LFSR();
double LCG();


int main() {
    static int flip[NUM_TOSS];
    printf("\n--------------------------\n");
    for (int j = 0; j < NUM_TOSS; j++) {
        flip[j] = findMode("LCG");
    }
    for (int i = 1; i < 26; i++) {
        printf("Length: %d num of heads: %d\n", i, numH(i, flip));
    }
    printf("\n--------------------------\n");
    for (int j = 0; j < NUM_TOSS; j++) {
        flip[j] = findMode("LFSR");
    }
    for (int i = 1; i < 26; i++) {
        printf("Length: %d num of heads: %d\n", i, numH(i, flip));
    }
    printf("\nEstimate Pi: LFSR\n");
    piEstimator("LFSR");
    printf("\nEstimate Pi: LCG\n");
    piEstimator("LCG");
    printf("\nRandom nums: LFSR\n");
    for (int i = 0 ; i < 10; i++) {
        printf("%f\n", LFSR());
    }
    printf("\nRandom nums: LCG\n");
    for (int i = 0 ; i < 10; i++) {
        printf("%f\n", LCG() );
    }
}
double LCG() {
    lcg = (A * lcg + C) % M;
    lcg = lcg % (int) pow(2,24);
    return lcg/pow(2,24);
}
double LFSR() {
    int bit23 = lfsr >> 23;
    int bit22 = lfsr >> 22;
    int bit21 = lfsr >> 21;
    int bit16 = lfsr >> 16;
    int XOR = bit23 ^ bit22 ^ bit21 ^ bit16;
    int Xand = XOR & 1;
    //System.out.println(Xand);
    lfsr = (lfsr << 1) | Xand;
    lfsr = lfsr % (int) pow(2,24);
    return lfsr/pow(2,24);

}
int findMode(char* ptr) {
    if (strcmp(ptr, "LFSR") == 0) {
        return (int) round(LFSR());
    }
    if (strcmp(ptr, "LCG") == 0) {
        return (int) round(LCG());
    }
    //need return
    return -1;
}
int numH(int length, int *flip) {
    int count = 0;
    int heads = 0;
    int boolF = 0;
    for (int i = 0; i < NUM_TOSS; i++) {
        if (flip[i]) {
            heads++;
            if (boolF) {
                boolF = 0;
            }
        } else {
            heads = 0;
            if (boolF) {
                count++;
                boolF = 0;
            }
        }
        if (heads == length) {
            boolF = 1;
            if (i == NUM_TOSS - 1) {
                count++;
                //System.out.println(count);
            }
        }
    }
    return count;
}
void piEstimator(char* ptr) {
    long z = 0, n = 0;
    double x, y, pi;
    for (int j = 0; j < 10; j++) {
        z = 0;
        n += 10000;
        for (long i = 0; i < n; i++) {
            if (strcmp(ptr, "LCG") == 0) {
                x = LCG();
                y = LCG();
            } else if (strcmp(ptr, "LFSR") == 0) {
                x = LFSR();
                y = LFSR();
            } else {
                return;
            }
            if (sqrt(x*x + y*y) <= 1) {
                z++;
            }
        }
        pi = 4.0*z/n;
        printf("Pi is estimated to be %f\n" , pi);
    }
}

File: LFSR.java
import java.util.*;
import java.io.*;
import java.math.*;


public class LFSR {
    public static void main (String[] args) {

	      LFSR(10);
	      LCG(10);
        pi_estimator();
    }


  public static double LFSR(int size) {
  int start = 12456723;
	int[] headCount = new int[25];
	System.out.println("LFSR random numbers:");
	double ans = 0;
	int count = 0;
	int countT = 0;
	for (int i = 0; i < size; i++) {
	    //System.out.println(Integer.toBinaryString(start));
	    int bit24 = start >> 24;
	    bit24 = bit24 & 1;
	    int bit23 = start >> 23;
	    bit23 = bit23 & 1;
	    int bit22 = start >> 22;
	    bit22 = bit22 & 1;
	    int bit17 = start >> 17;
	    bit17 = bit17 & 1;
	    int XOR = bit24 ^ bit23 ^ bit22 ^ bit17;
	    start = start << 1 ^ XOR;
	    start = start % (int) Math.pow(2,24);
	    ans = start/Math.pow(2,24);
      // double newAns = ans * 10000000;
      // newAns = Math.floor(newAns);
      // start = (int) newAns;
	    //System.out.println(Integer.toBinaryString(start));
	    System.out.println(ans);

	    if ( ans > .5) {
		      count++;
		      countT = 0;
	    } else {
		      countT++;
		        if (countT < 2 && count > 0) {
		            headCount[count - 1] = headCount[count - 1] + 1;
		            count = 0;
		        }
	    }
	  }
	if (count > 0) {
	    headCount[count - 1] = headCount[count - 1] + 1;

	}

	for (int s = 0; s < 25; s++) {
	    System.out.print(headCount[s] + " ");
	}
  System.out.println(ans);
  return 0.0;
	//System.out.println();
    }

  public static void LCG(int size) {
	System.out.println("LCG random numbers:");
	int start = 12456723;
	double ans;
	int[] headCount = new int[25];
	int count = 0;
	int countT = 0;
	for(int i = 0; i < size; i++) {
	    start =(int) (1140671485*start + 128201163) % (int) Math.pow(2,24);
	    ans = Math.abs(start / Math.pow(2,24));
      // double newAns = ans * 10000000;
      // newAns = Math.floor(newAns);
      // start = (int) newAns;
	    if ( ans > .5) {
		      count++;
		      countT = 0;
		      //System.out.println("H");
	    } else {
		      //System.out.println("T");
		      countT++;
		      if (countT < 2 && count > 0) {
		          headCount[count - 1] = headCount[count - 1] + 1;
		            count = 0;
		      }
	    }
	    System.out.println(ans);
	}
	if (count > 0) {
	    headCount[count - 1] = headCount[count - 1] + 1;
	}
	for (int s = 0; s < 25; s++) {
	    System.out.print(headCount[s] + " ");
	}
	//System.out.println();

    }


    public static void pi_estimator() {
        long z=0, n=1000000;
        double x, y, pi;

        Random rnd = new Random();

        for (long i=0; i < n; ++i) {
          x = rnd.nextDouble();
          y = rnd.nextDouble();
          if (Math.sqrt(x*x + y*y) <= 1) z++;
        }

        pi = 4.0*z/n;
        System.out.println("Pi is estimated to be " + pi);


    }
}
